saleor/graphql/order/mutations/fulfillments.py

```diff
@@ -308,3 +312,245 @@ def perform_mutation(cls, _root, info, **data):
+
+        if not order_lines_data and not fulfillment_lines_data:
+            raise ValidationError(
+                {
+                    "order_lines": ValidationError(
+                        "The orderLines or fulfillmentLines cannot be empty.",
+                        code=OrderErrorCode.REQUIRED.value,
+                    ),
+                    "fulfillment_lines": ValidationError(
+                        "The orderLines or fulfillmentLines cannot be empty.",
+                        code=OrderErrorCode.REQUIRED.value,
+                    ),
+                }
+            )
+        if order_lines_data:
+            cls.clean_lines(order_lines_data, cleaned_input)
+        if fulfillment_lines_data:
+            cls.clean_fulfillment_lines(fulfillment_lines_data, cleaned_input)
+        return cleaned_input
+
+    @classmethod
+    def _raise_error_for_line(cls, msg, type, line_id, field_name, code=None):
+        line_global_id = graphene.Node.to_global_id(type, line_id)
+        if not code:
+            code = OrderErrorCode.INVALID_REFUND_QUANTITY.value
+        raise ValidationError(
+            {
+                field_name: ValidationError(
+                    msg, code=code, params={field_name: line_global_id},
+                )
+            }
+        )
+
+    @classmethod
+    def clean_fulfillment_lines(cls, fulfillment_lines_data, cleaned_input):
+        lines_ids = [line["fulfillment_line_id"] for line in fulfillment_lines_data]
+        quantities_to_refund = [line["quantity"] for line in fulfillment_lines_data]
+        lines_to_refund = cls.get_nodes_or_error(
+            lines_ids,
+            field="fulfillment_lines",
+            only_type=FulfillmentLine,
+            qs=order_models.FulfillmentLine.objects.prefetch_related("fulfillment"),
+        )
+        lines_to_refund = list(lines_to_refund)
+        for line, quantity in zip(lines_to_refund, quantities_to_refund):
+            if line.quantity < quantity:
+                cls._raise_error_for_line(
+                    "Quantity provided to refund is bigger than quantity from "
+                    "fulfillment line",
+                    "FulfillmentLine",
+                    line.pk,
+                    "fulfillment_line_id",
+                )
+            if line.fulfillment.status != FulfillmentStatus.FULFILLED:
+                cls._raise_error_for_line(
+                    "Unable to refund fulfillmentLine with different type than "
+                    "FULFILLED.",
+                    "FulfillmentLine",
+                    line.pk,
+                    "fulfillment_line_id",
+                    code=OrderErrorCode.CANNOT_REFUND_FULFILLMENT_LINE.value,
+                )
+        cleaned_input["fulfillment_lines_quantities_to_refund"] = quantities_to_refund
+        cleaned_input["fulfillment_lines_to_refund"] = lines_to_refund
+
+    @classmethod
+    def clean_lines(cls, lines_data, cleaned_input):
```

Comment:
`clean_lines` looks similar to `clean_fulfillment_lines` cannot create one generic function?

Reply:
I am afraid I can not, or at least fast and clean. The conditions and the error msg for fulfillmentLines and orderLines are different.

Reply:
👍